x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseRunAsSuperuserCommand.java

```diff
@@ -0,0 +1,281 @@
+    /**
+     * Removes temporary file realm user from users and roles file
+     */
+    protected void cleanup(Terminal terminal, Environment env) throws Exception {
+        final Path passwordFile = FileUserPasswdStore.resolveFile(env);
+        final Path rolesFile = FileUserRolesStore.resolveFile(env);
+        FileAttributesChecker attributesChecker = new FileAttributesChecker(passwordFile, rolesFile);
+
+        Map<String, char[]> users = FileUserPasswdStore.parseFile(passwordFile, null, env.settings());
+        if (users == null) {
+            throw new UserException(ExitCodes.CONFIG, "File realm configuration file [" + passwordFile + "] is missing");
+        }
+        users = new HashMap<>(users);
+        char[] passwd = users.remove(username);
+        if (passwd != null) {
+            // No need to overwrite, if the user was already removed
+            FileUserPasswdStore.writeFile(users, passwordFile);
+            Arrays.fill(passwd, '\0');
+        }
+
+        Map<String, String[]> userRoles = FileUserRolesStore.parseFile(rolesFile, null);
+        if (userRoles == null) {
+            throw new UserException(ExitCodes.CONFIG, "File realm configuration file [" + rolesFile + "] is missing");
+        }
+        userRoles = new HashMap<>(userRoles);
+        String[] roles = userRoles.remove(username);
+        if (roles != null) {
+            // No need to overwrite, if the user was already removed
+            FileUserRolesStore.writeFile(userRoles, rolesFile);
+        }
+        attributesChecker.check(terminal);
+    }
+
+    protected SecureString getPassword() {
+        return password;
+    }
+
+    protected String getUsername() {
+        return username;
+    }
+
+    private void ensureFileRealmEnabled(Settings settings) throws Exception {
+        Map<String, Settings> fileRealmSettings = settings.getGroups("xpack.security.authc.realms.file");
+        if (fileRealmSettings.size() > 1) {
+            throw new UserException(
+                ExitCodes.CONFIG,
+                "Multiple file realms are configured. "
+                    + "[file] is an internal realm type and therefore there can only be one such realm configured"
+            );
+        } else if (fileRealmSettings.size() == 1
+            && fileRealmSettings.entrySet().stream().anyMatch(s -> s.getValue().get("enabled").equals("false"))) {
+                throw new UserException(ExitCodes.CONFIG, "File realm must be enabled");
+            }
+        // Else it's either explicitly enabled, or not defined in the settings so it is implicitly enabled.
+    }
+
+    /**
+     * Checks that we can connect to the cluster and that the cluster health is not RED. It optionally handles
+     * retries as the file realm might not have reloaded the users file yet in order to authenticate our
+     * newly created file realm user.
+     */
+    private void checkClusterHealthWithRetries(Environment env, Terminal terminal, int retries) throws Exception {
```

Comment:
We were chatting around this with @albertzaharovits today. This seems reasonable as a base check but were wondering whether a more nuanced approach is necessary/helpful. 

The concern is that for some tools that extend this command and have specialized purpose and needs, this check might be overly coarse grained. Point in case, for the reset password tool, we might not be interested if any data index is red( making the cluster health red ), as long as `.security` is green. On the other hand, we want this to be useful even if the .security index is not yet created so we'd need to check for the existence of the .security index before the health check so that it won't time out. We can't use the exists API in 8.0 for system indices, so we'd have to use `_cat/indices` or `_cat/aliases` and parse the response. 

The question is whether we need this nuanced approach, and if we do, do we implement it here in the base class or do we abstract it and let the extending classes implement this as they see fit. Given in the two classes that extend this (CreateEnrollmentTokenTool and ResetElasticPasswordTool) so far, we only care about the security index, I lean on doing the former.

Reply:
` _cat/indices` and/or ` _cat/aliases` are not that friendly for parsing the output. Another idea would be to not attempt to do the health check up front, but depend on the error message that we would get from Elasticsearch in order to determine if the security index health is RED. There are a couple of issues with this too:
- We don't consistently check the health of the security index in the SecurityIndexManager before making operations on it. Some methods do manual checks like `isAvailable()` and then call `checkIndexVersionThenExecute`, some call `prepareIndexIfNeededThenExecute()` directly which doesn't call `isAvailable()`. So we would need to depend on the behavior of the actual indexing of the document in the index. 
- Not all implementing classes of BaseRunAsSuperuser will be calling APIs with the CommandLineHttpClient themselves. Reset password does, but CreateEnrollmentTokenTool does not, instead depends on CreateEnrollmentToken methods that call the APIs internally and will wrap exceptions. 

It should still be doable but some of it sounds like guesswork.

Is "if your cluster health is RED, you shouldn't be making additional operations on it (like using these tools) unless it's about resolving the issue, so the tool doesn't work if cluster health is RED" an appropriate stance or do we think that a nuanced approach ( i.e. each tool defines it's own health/index checks as it sees fit and the base class just validates file realm auto-generated credentials ) is helpful and necessary ?

Opening up the discussion, @tvernum I know from previous discussions that you have opinions on this type of behavior by our CLI tools, care to chime in?

Reply:
_Disclaimer:_ Speaking in principle, haven't checked the code too carefully:
Any check that we do before the actual operation has limited usefulness because the condition can change in between the time of check and the operation time (time of check, time of use).
I think it's worth investing in some tests that "pin" the errors (ideally consistent across APIs) for an unavailable/red .security index.

Reply:
We chatted with Tim a little bit about this and we came to a similar conclusion which was that until we can identify all the ways this can fail in an actionable and informative way, replicating the behavior of other tools ( warn if cluster health is red and allow to override ) is preferable to complicate the checks for handling _some_ of the cases. I plan on moving forward as is and open an issue to track the effort to do what you suggest, if you don't think this is blocking for the current PR